src/par_iter/find_first_last.rs

```diff
@@ -0,0 +1,319 @@
+            find_op: find_op,
+            lower_bound: Cell::new(0),
+            upper_bound: usize::max_value(),
+            match_position: match_position,
+            best_found: best_found,
+        }
+    }
+
+    fn current_index(&self) -> usize {
+        match self.match_position {
+            MatchPosition::Leftmost => self.lower_bound.get(),
+            MatchPosition::Rightmost => self.upper_bound
+        }
+    }
+}
+
+impl<'f, ITEM, FIND_OP> Consumer<ITEM> for FindConsumer<'f, FIND_OP>
+    where ITEM: Send,
+          FIND_OP: Fn(&ITEM) -> bool + Sync
+{
+    type Folder = FindFolder<'f, ITEM, FIND_OP>;
+    type Reducer = FindReducer;
+    type Result = Option<ITEM>;
+
+    fn cost(&mut self, cost: f64) -> f64 {
+        cost * FUNC_ADJUSTMENT
+    }
+
+    fn split_at(self, _index: usize) -> (Self, Self, Self::Reducer) {
+        let dir = self.match_position;
+        (self.split_off(),
+         self,
+         FindReducer { match_position: dir })
+    }
+
+    fn into_folder(self) -> Self::Folder {
+        FindFolder {
+            find_op: self.find_op,
+            boundary: self.current_index(),
+            match_position: self.match_position,
+            best_found: self.best_found,
+            item: None,
+        }
+    }
+
+    fn full(&self) -> bool {
+        // can stop consuming if the best found index so far is *strictly*
+        // better than anything this consumer will find
+        better_position(self.best_found.load(Ordering::Relaxed),
+                        self.current_index(),
+                        self.match_position)
+    }
+}
+
+#[test]
+fn same_range_consumers_return_correct_answer() {
+    let find_op = |x: &i32| x % 2 == 0;
+
+    let first_found = AtomicUsize::new(usize::max_value());
+    let first_consumer = FindConsumer::new(&find_op,
+                                           MatchPosition::Leftmost,
+                                           &first_found);
+
+    // split until we have an indivisible range
+    let bits_in_usize = usize::min_value().count_zeros();
+    for i in 0..bits_in_usize {
+        first_consumer.split_off();
+    }
+
+    let first_reducer = first_consumer.to_reducer();
+    // the left and right folders should now have the same range, having
+    // exhausted the resolution of usize
+    let left_first_folder = first_consumer.split_off().into_folder();
+    let right_first_folder = first_consumer.into_folder();
+
+    let left_first_folder = left_first_folder.consume(0).consume(1);
+    assert_eq!(left_first_folder.boundary, right_first_folder.boundary);
+    // expect not full even though a better match has been found because the
+    // ranges are the same
+    assert!(!right_first_folder.full());
+    let right_first_folder = right_first_folder.consume(2).consume(3);
+    assert_eq!(first_reducer.reduce(left_first_folder.complete(),
+                                    right_first_folder.complete()),
+               Some(0));
```

Comment:
Do you mean that you want to see a consumer from outside of the exhausted range, so that we can check that it's full when the consumer from inside the range finds a match? That wasn't quite what your example code does (you'd have to move the creation of `second_folder` above the loop, or something like that), but I think that's what you're going for.